bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/ClientLibrariesImpl.java

```diff
@@ -0,0 +1,297 @@
+                }
+            }
+        } catch (IOException e) {
+            LOG.error("Failed to include client libraries {}", categoriesArray);
+        }
+
+        String html = sw.toString();
+        // inject attributes from HTL into the JS and CSS HTML tags
+        return getHtmlWithInjectedAttributes(html);
+    }
+
+    private String getHtmlWithInjectedAttributes(String html) {
+        StringBuilder jsAttributes = new StringBuilder();
+        jsAttributes.append(getHtmlAttr(OPTION_ASYNC, async));
+        jsAttributes.append(getHtmlAttr(OPTION_DEFER, defer));
+        jsAttributes.append(getHtmlAttr(OPTION_CROSSORIGIN, crossorigin));
+        jsAttributes.append(getHtmlAttr(OPTION_ONLOAD, onload));
+        StringBuilder cssAttributes = new StringBuilder();
+        cssAttributes.append(getHtmlAttr(OPTION_MEDIA, media));
+        String updatedHtml = StringUtils.replace(html,"<script ", "<script " + jsAttributes.toString());
+        return StringUtils.replace(updatedHtml,"<link ", "<link " + cssAttributes.toString());
+    }
+
+    private String getHtmlAttr(String name, boolean include) {
+        if (include) {
+            return name + " ";
+        }
+        return "";
+    }
+
+    private String getHtmlAttr(String name, String value) {
+        if (StringUtils.isNotEmpty(value)) {
+            return name + "=\"" + value + "\" ";
+        }
+        return "";
+    }
+
+    private String getInline(LibraryType libraryType) {
+        Collection<ClientLibrary> clientlibs = htmlLibraryManager.getLibraries(categoriesArray, libraryType, true, false);
+        // Iterate through the clientlibs and aggregate their content.
+        StringBuilder output = new StringBuilder();
+        for (ClientLibrary clientlib : clientlibs) {
+            HtmlLibrary htmlLibrary = htmlLibraryManager.getLibrary(libraryType, clientlib.getPath());
+            if (htmlLibrary != null) {
+                try {
+                    output.append(IOUtils.toString(htmlLibrary.getInputStream(htmlLibraryManager.isMinifyEnabled()),
+                        StandardCharsets.UTF_8));
+                } catch (IOException e) {
+                    LOG.error("Error getting input stream from clientlib with path '{}'.", clientlib.getPath());
+                }
+            }
+        }
+        return output.toString();
+    }
+
+    public Set<String> getCategoriesFromComponents() {
+        Set<String> categories = new HashSet<>();
+
+        allLibraries = htmlLibraryManager.getLibraries();
+        Collection<ClientLibrary> libraries = new LinkedList<>();
+
+        for (String resourceType : resourceTypes) {
+            Resource componentRes = getResource(resourceType);
+            addClientLibraries(componentRes, libraries);
+
+            if (inherited && componentRes != null) {
+                addClientLibraries(getResource(componentRes.getResourceSuperType()), libraries);
```

Comment:
This only adds the super type of the resource (one level inheritance). How about looking up the whole inheritance chain (multi level inheritance)?

Reply:
Good catch! This used to be called recursively before some of the refactoring. Will need to move the inherited check to `addClientLibraries()`

Reply:
Done in #1100